metadata/metadata.go

```diff
@@ -113,26 +113,56 @@ type mdOutgoingKey struct{}
 
 // NewIncomingContext creates a new context with incoming md attached.
 func NewIncomingContext(ctx context.Context, md MD) context.Context {
-	return context.WithValue(ctx, mdIncomingKey{}, md)
+	return context.WithValue(ctx, mdIncomingKey{}, []MD{md})
 }
 
 // NewOutgoingContext creates a new context with outgoing md attached.
 func NewOutgoingContext(ctx context.Context, md MD) context.Context {
-	return context.WithValue(ctx, mdOutgoingKey{}, md)
+	return context.WithValue(ctx, mdOutgoingKey{}, []MD{md})
+}
+
+// AppendToOutgoingContext returns a new context with outgoing md attached
+func AppendToOutgoingContext(ctx context.Context, md MD) context.Context {
+	mds, _ := ctx.Value(mdOutgoingKey{}).([]MD)
```

Comment:
This will panic if no MD has been attached yet, I believe.

Reply:
I tested locally before and it seemed fine. It evaluates to an empty array.

Reply:
Hmm, I'm seeing something different; I wonder why?

https://play.golang.org/p/F7q5wgSB-Q0

Reply:
@dfawley Ok, but are you ready for something really spooky? https://play.golang.org/ 😆

Reply:
> Ah shoot you already got it, sorry haha. I should have refreshed page. Yeah really strange behavior

It's a bit odd, but this is a feature of type assertions.  An untyped nil doesn't type-assert to anything, and a type assertion without a "`, ok`" will panic if the type assertion fails.